Sonos responsiveness improvements + enhancements

bjarniivarsson opened the pull request:
**Description:**
- Event based notification directly from the speakers.
- Async state updates.
- Optimizations to minimize chattyness with speakers.
- Preloading/caching of album art for next song.
- Extracting "Information" or "Now" data for radio streams (show as the secondary text in the UI).
- Disable previous/next track actions for first/last item in a playlist/queue.
- Event based volume/mute changes from speaker, so that volume_level and is_volume_muted don't need to do network IO.
- More robust method of detecting when a radio stream is being played.
- Detecting title for next item in playlist/queue (this is currently not exposed, will require changes to media_player and UI).

This code applies a patch to SoCo automatically and works around another SoCo issue, I have submitted a couple of PR's to SoCo (the code will work without them):
- https://github.com/SoCo/SoCo/pull/437
- https://github.com/SoCo/SoCo/pull/438

**Checklist:**

If the code communicates with devices, web services, or third-party tools:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

bjarniivarsson committed:
Sonos responsiveness improvements (async_ coroutines, event based updating, album art caching) + Better radio station information

bjarniivarsson committed:
Docstring fixes.

bjarniivarsson committed:
Docstring fixes.

pvizeli reviewed (dismissed):
I like the refactory of sonos component. But some things is not conform. I don't like the image cache. The frontent don't read a lot of the picture value. But I don't know what @balloob thinking about this.

You should not use asyncio while soco is a blocking library. In this case hass core while execute all thing in a executor and you don't need to do that. 1-2 executor calls are okay but you have a lot of them and overwrite some things on Mediaplayer entity.

I think also that we should not mock the soco library. Please add first you patch to soco and after it give a new release you can add this to Home-Assistant. Our developer guideline say that we not merge device specifiy code and they should be in a own library.

bjarniivarsson commented:
@pvizeli Thanks for the review. Some points:

- The image cache is there to reduce the latency between selecting the next track in a playlist (pressing the next button) and the album art appearing. During my experimentation I found that most of the latency was being spent on the request from HA to the speaker. This becomes more extreme if you have several instances of the same album art being display (e.g. when grouping speakers, or running multiple UI's). I can however remove this from this PR and make another one later so that we can discuss that separately. Would that be acceptable or do we want input from others on this as well?

- Regarding the asyncio stuff, most of the updating is being done based on the data in the events from SoCo (the event handler is a coroutine), and my plan was to increase that further, so that the entity could be fully driven (asynchronously) by the events. In the meantime it might make sense to replace the SoCo wrapping async methods with one or two methods that do all of the work at once, would that be acceptable?

- The SoCo PR's have been accepted and merged, I'm not sure about the release schedule there. Not sure what the policy is in these things, Realize that we don't want to implement device specific code in HA, but one could argue that most of the code in sonos.py is device specific. Is it not acceptable to have workarounds for problems in device specific code? Or do we always want to wait for fixes to trickle out?

robbiet480 commented:
😍 . Thanks so much for this, it looks great. 


As for the requirements and the SoCo release scheduled etc. you can just point to a specific Git commit in the requirements line. For example, here's the GTFS sensor requirements:

```python
REQUIREMENTS = ["https://github.com/robbiet480/pygtfs/archive/"
                "00546724e4bbcb3053110d844ca44e2246267dd8.zip#"
                "pygtfs==0.1.3"]
```

Even your own fork of SoCo is acceptable. The only rule is *it must be a specific commit, not a branch*. 

I also really don't like the patching. Would rather see you make the chance on your fork and point requirements to that. As @pvizeli said, the mocking is also a no go since it is device specific code.

Looking forward to seeing this in 0.32!

balloob commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -17,12 +19,34 @@
+_SOCO_PATCHED = False
+
+
+def _patch_soco():
```
I don't like this. I see that your patch has been merged so I rather have them put out a new release or we start depending on a specific commit in their repo.

balloob commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -5,6 +5,7 @@
 https://home-assistant.io/components/media_player/
```
You changed file permissions.

balloob commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -674,6 +675,31 @@ def state_attributes(self):
+        )
+
+
+@functools.lru_cache(maxsize=16)
```
I like it 👍

lwis commented:
This looks great, this will improve the reliability of a few of my automations massively.

bjarniivarsson committed:
Updated SoCo dependency + fixed file permissions.

bjarniivarsson committed:
Only fetch speaker info if needed.

bjarniivarsson committed:
PEP8 fixes

bjarniivarsson commented:
Hi,

Made some changes:

- Removed SoCo patching and updated dependency to point to my SoCo commit. I asked them about a new release, but since they just released it will not be immediate.
- Fixed the file permission changes.
- Now we only fetch speaker_info if we don't already have it. We were fetching it for every call to async_update before, but were not getting the latest info anyway (SoCo caches it). Now we fetch it and force SoCo to go over the network initially. This basically means that we will not detect any changes to the speaker name (which is unchanged behavior).
- I confirmed, that when receiving events from the speaker we only ever make one "async wrapped" call to SoCo (_async_get_media_info in line 480), all other data is pulled from the event. The update method is still being polled by HA, and I think we should keep it that way for now, it is useful if the event subscriptions from the speaker break down (which to be fair, I have not seen happening).

bjarniivarsson committed:
Fixed SoCoMock.get_speaker_info to get test to pass.

bjarniivarsson commented:
Also, I did not remove the image caching since @balloob seems to like it, I could pull it out and put that into a separate PR later if you wish.

lwis commented:
@bjarniivarsson my view is that it's a positive change unless we have problems.

bjarniivarsson commented:
I have no idea why the checks are failing, does not seem related to my changes.

lwis commented:
@bjarniivarsson test looks fine, could you re-run `script/gen_requirements_all.py` to see if it yields any changes?

Also the coverage has decreased by .1%, this can't be merged until that's resolved.

pvizeli commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -674,6 +675,31 @@ def state_attributes(self):
+    size).
+    """
+    try:
+        response = requests.get(url)
```
That will run inside async loop, so you need to use aiohttp instead request with hass.websession as SessionClient

lwis replied:
@pvizeli I don't think `lru_cache` works with coroutines. We may need to write our own, something like; http://stackoverflow.com/a/34123396. We should also consider using `wrapt` (https://hynek.me/articles/decorators/)

bjarniivarsson replied:
@lwis I was not sure about that either, and therefore I put it on a method that is not a coroutine, don't see why it would not work.

lwis replied:
Will this function be called from within the event loop?

bjarniivarsson replied:
I get where you're coming from :) Yes this it will be called from within the event loop, and if I change it to use aiohttp it will need to be a coroutine and that it would stop working, so you're right :). Will have a look at another approach.

robbiet480 commented:
@lwis @bjarniivarsson Coverage decreasing by -0.1% is within tolerance levels. Coveralls is a bit hit or miss sometimes like in this situation.

bjarniivarsson committed:
Regenerated requirements_all.txt + async fetching of album art with caching + added http_session to HomeAssistant object.

bjarniivarsson commented:
@lwis @pvizeli Implemented async loading of album art with a very simple (no new dependencies) caching solution. I had some problems deciding where to get the aiohttp.ClientSession to use for the loading and ended up putting it on the HomeAssistand (hass) object. The documentation encourages that application use a single global ClientSession instance for all communication. I could put it somewhere else (even create a new one every time, but then we loose the keep-alive functionality).

bjarniivarsson commented:
Someone beat me to the global ClientSession :)

bjarniivarsson committed:
Unit test fixed.

lwis reviewed (approved):
I'd have liked to keep the LRUCache, but as a first pass this looks good.

bjarniivarsson commented:
@lwis thanks for approving :) I wanted to keep it simple for now, I'll have a look at alternatives and if I see something good, I'll make another PR for that.

robbiet480 committed:
Add blank line as per flake8

bjarniivarsson committed:
Fixed media image proxy unit test.

robbiet480 commented:
Tests are failing due to flakiness. Other than that, looks good to merge if everyone is cool. Looks good to me!

balloob commented:
So, want to chime in here, I'm a bit late, I know that a lot of work has already been done. However, I don't think that it has any benefit to use any async stuff if the underlying lib is not async. In that case we will run everything in the executor and by not migrating anything, Home Assistant will run the whole update sequence in the executor - probably faster overall.

robbiet480 commented:
A lot of this PR is usable and good though.

> * Optimizations to minimize chattyness with speakers.
* Preloading/caching of album art for next song.
* Extracting "Information" or "Now" data for radio streams (show as the secondary text in the UI).
* Disable previous/next track actions for first/last item in a playlist/queue.
* Event based volume/mute changes from speaker, so that volume_level and is_volume_muted don't need to do network IO.
* More robust method of detecting when a radio stream is being played.
* Detecting title for next item in playlist/queue (this is currently not exposed, will require changes to media_player and UI).

bjarniivarsson commented:
Thanks again for looking at this PR. If the async stuff is prevent this from being accepted, I'll go ahead and pull that out, probably should not have done that along with all the functionality improvements. Let me know if I should do it :)

pvizeli commented:
It is a greate PR anyway. Thanks for you hard work I like to use on my sonos 👍

bjarniivarsson committed:
Removed async stuff.

bjarniivarsson commented:
@balloob @robbiet480 Removed the async stuff, and again there are unrelated (to me) tests failing :(

robbiet480 commented:
Tests are failing due to unrelated flaky tests, this is good to merge.

pvizeli commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -223,6 +225,29 @@ def wrapper(*args, **kwargs):
+
+
+# pylint: disable=too-few-public-methods
+class _AsyncProcessSonosEventQueue():
```
You should rename that.

pvizeli reviewed (commented):
A small comment to need address and if all test are okay and coverage don't be bad, I can give a 🐬

bjarniivarsson committed:
Removed last remnants of async stuff.

bjarniivarsson commented:
I think this is it, one glitchy test failing.

robbiet480 commented:
So happy that I get to be the one to hit merge on this 😺

balloob commented:
This PR introduced some serious test slowness for Sonos. It's printed at the end of every test run on CI. After this PR:

```
========================== slowest 10 test durations ===========================
3.14s call     tests/components/media_player/test_sonos.py::TestSonosMediaPlayer::test_sonos_unjoin
3.13s call     tests/components/media_player/test_sonos.py::TestSonosMediaPlayer::test_ensure_setup_discovery
3.10s call     tests/components/media_player/test_sonos.py::TestSonosMediaPlayer::test_sonos_restore
3.10s call     tests/components/media_player/test_sonos.py::TestSonosMediaPlayer::test_sonos_set_sleep_timer
3.09s call     tests/components/media_player/test_sonos.py::TestSonosMediaPlayer::test_sonos_group_players
3.09s call     tests/components/media_player/test_sonos.py::TestSonosMediaPlayer::test_sonos_clear_sleep_timer
3.09s call     tests/components/media_player/test_sonos.py::TestSonosMediaPlayer::test_sonos_snapshot
3.09s call     tests/components/media_player/test_sonos.py::TestSonosMediaPlayer::test_ensure_setup_sonos_discovery
3.09s call     tests/components/media_player/test_sonos.py::TestSonosMediaPlayer::test_ensure_setup_config
```

Could you address this @bjarniivarsson ?

bjarniivarsson commented:
Absolutely @balloob, I'll have a look, should have some free time over the weekend.

lwis commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -268,76 +305,140 @@ def name(self):
+
+    def _is_available(self):
+        try:
+            sock = socket.create_connection(
```
@bjarniivarsson we should mock this in tests.
